text/0957-modernized-scheduler.md

```diff
@@ -0,0 +1,633 @@
+overhead light, and enables async stack traces for scheduled work to maintain the context
+of where the work was scheduled.
+
+For instance, take the partial strategy implementation shown below which handles `render`
+`layout` and `composite` by scheduling three `requestAnimationFrame` callbacks wrapped in
+promises and allows recursive calls to render and just-in-time calls to layout and composite.
+
+**Example 1**
+```ts
+class Scheduler {
+  _nextRender = null;
+  _nextLayout = null;
+  _nextComposite = null;
+  _isFlushing = false;
+  _isFlushingRender = false;
+  _flushComplete = null;
+
+  ensureFrame(name) {
+    if (!this._nextRender) {
+      if (this._isFlushingRender) {
+          this._nextRender = Promise.resolve();
+      } else if (!this._isFlushing || name === 'render') {
+        this._nextRender = new Promise(resolve => {
+          requestAnimationFrame(() => {
+            console.log('flushing render');
+            this._isFlushing = true;
+            this._isFlushingRender = true;
+            this._nextRender = null;
+            resolve();
+          });
+        });
+      }
+    }
+
+    if (!this._nextLayout) {
+      if (!this._isFlushing || name === 'layout') {
+        this._nextLayout = new Promise(resolve => {
+          requestAnimationFrame(() => {
+            console.log('flushing layout');
+            this._isFlushingRender = false;
+            this._nextRender = null;
+            this._nextLayout = null;
+            resolve();
+          });
+        });
+      }
+    }
+
+    if (!this._nextComposite) {
+      this._nextComposite = new Promise(resolve => {
+        requestAnimationFrame(() => {
+          console.log('flushing composite');
+          this._nextComposite = null;
+          resolve();
+        });
+      });
+    }
+
+    if (!this._flushComplete) {
+      this._flushComplete = new Promise(resolve => {
+        requestAnimationFrame(() => {
+          console.log('flushing complete');
+          this._isFlushingRender = false;
+          this._isFlushing = false;
+          this._flushComplete = null;
+          resolve();
+        });
+      });
+    }
+  }
+
+  render() {
+      this.ensureFrame('render');
+      return this._nextRender;
+  }
+
+  layout() {
+      this.ensureFrame('layout');
+      return this._nextLayout;
+  }
+
+  composite() {
+      this.ensureFrame('composite');
+      return this._nextComposite;
+  }
+}
+const scheduler = new Scheduler();
+```
+
+Using this strategy, lets observe what happens when we use normal `async/await`
+patterns to schedule some work.
+
+**Example 2**
+```ts
+
+// puts some blocking named work into the performance profiler
+// to make it more obvious where the work was done and what work it was
+function doExpensiveWork(name, step) {
+  const fn = new Function(`return function ${name}() {const start = performance.now(); while (performance.now() - start < 10) {} console.log('${step}. ${name}');};`);
```

Comment:
```suggestion
  const fn = new Function(`
    return function ${name}() {
      const start = performance.now(); 
      while (performance.now() - start < 10) {} 
      console.log('${step}. ${name}');
    };`
  );
```

suggestion to avoid horizontal scrolling